backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/cenum/CEnumByValueFunctionGenerator.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package org.jetbrains.kotlin.backend.konan.ir.interop.cenum
+
+import org.jetbrains.kotlin.backend.konan.ir.interop.DescriptorToIrTranslationMixin
+import org.jetbrains.kotlin.backend.konan.ir.interop.findDeclarationByName
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.descriptors.IrBuiltIns
+import org.jetbrains.kotlin.ir.symbols.IrSimpleFunctionSymbol
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.isInt
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.util.irCall
+import org.jetbrains.kotlin.util.OperatorNameConventions
+
+/**
+ * Generate IR for function that returns appropriate enum entry for the provided integral value.
+ */
+internal class CEnumByValueFunctionGenerator(
+        override val irBuiltIns: IrBuiltIns,
+        override val symbolTable: SymbolTable,
+        override val typeTranslator: TypeTranslator
+) : DescriptorToIrTranslationMixin {
+    fun generateByValueFunction(
+            companionIrClass: IrClass,
+            valuesIrFunctionSymbol: IrSimpleFunctionSymbol
+    ): IrFunction {
+        val byValueFunctionDescriptor = companionIrClass.descriptor.findDeclarationByName<FunctionDescriptor>("byValue")!!
+        val byValueIrFunction = declareSimpleIrFunction(byValueFunctionDescriptor)
+        val irValueParameter = byValueIrFunction.valueParameters.first()
+        // checkNotNull {
+        // val values: Array<E> = values()
+        // val i: Int = 0
+        // val size: Int = values.size
+        // while (i < size) {
+        //      val entry: E = values[i]
+        //      val entryValue = entry.value
+        //      if (entryValue == arg) {
+        //          return entry
+        //      }
+        //      i++
+        // }
+        // return null
+        // }
+        byValueIrFunction.body = irBuilder(irBuiltIns, byValueIrFunction.symbol).irBlockBody {
+            +irReturn(irForceNotNull(irBlock {
```

Comment:
`irForceNotNull` is kinda useless here, since `result == null` is known for each `irReturn`.
It might be easier to call `symbols.ThrowNullPointerException` explicitly.